plugins/bcc-login/includes/class-bcc-login-visibility.php

```diff
@@ -192,8 +194,91 @@ function on_template_redirect() {
                 return $this->not_allowed_to_view_page($visited_url);
             }
         }
+    }
+
+    /**
+     * Fail json request if the post requires a higher level.
+     *
+     * @return array
+     */
+    function on_rest_pre_echo_response( $response, $object, $request ) {
+        $route = $request->get_route();
+        $session_is_valid = $this->_client->is_session_valid();
+        $user_level = (int) $this->_client->get_user_level_based_on_claims();
+        $user_groups = $this->get_current_user_groups();
+
+        if ($route == '/wp/v2/search') {
+            $response_arr = [];
+
+            if (array_key_exists('code', $response) && $response['code'] == 'rest_cookie_invalid_nonce')
+                return $response;
+
+            foreach ($response as $item) {
+                $post_visibility = (int) get_post_meta( $item['id'], 'bcc_login_visibility', true );
+
+                if ($session_is_valid) {
+                    // Check login visibility
+                    if ( $post_visibility > $user_level ) {
+                        continue;
+                    }
+
+                    // Check user groups
+                    if ( !empty($this->_settings->site_groups) && !current_user_can( 'edit_posts' ) ) {
+                        $post_groups = get_post_meta( $item['id'] , 'bcc_groups', false );
+    
+                        if ($post_groups && !$user_groups) {
+                            continue;
+                        }
+            
+                        if (count(array_intersect($post_groups, $user_groups)) == 0 &&
+                            count(array_intersect($this->_settings->full_content_access_groups, $user_groups)) == 0)
+                        {
+                            continue;
+                        }
+                    }
+
+                    $response_arr[] = $item;
+                }
+                else if ($post_visibility <= (int) $this->_settings->default_visibility) {
+                    $response_arr[] = $item;
+                }
+            }
+
+            return $response_arr;
+        }
 
+        else if (preg_match('#^/wp/v2/(' . implode('|', $this->visibility_post_types) . ')/(\d+)$#', $route, $matches) ) {
+            $post_visibility = (int) $response['meta']['bcc_login_visibility'];
 
+            if ($session_is_valid) {
+                // Check login visibility
+                if ( $post_visibility > $user_level ) {
+                    return $this->not_allowed_to_view_page();
+                }
+    
+                // Check user groups
+                if ( !empty($this->_settings->site_groups) && !current_user_can( 'edit_posts' ) ) {
```

Comment:
This code seems to be duplicated here (and probably elsewhere in the plugin). Perhaps we should move it to a function which takes post and user as arguments or something?

Reply:
The code looks duplicated in the 3 instances but it is actually quite different. I don't see it necessary or doable.